scio-core/src/main/scala/com/spotify/scio/coders/CoderMaterializer.scala

```diff
@@ -17,20 +17,91 @@
 
 package com.spotify.scio.coders
 
-import org.apache.beam.sdk.coders.{Coder => BCoder, IterableCoder, KvCoder, NullableCoder}
+import com.spotify.scio.util.RemoteFileUtil
+import org.apache.beam.sdk.coders.{
+  Coder => BCoder,
+  IterableCoder,
+  KvCoder,
+  NullableCoder,
+  ZstdCoder
+}
 import org.apache.beam.sdk.options.{PipelineOptions, PipelineOptionsFactory}
 
+import java.net.URI
+import java.nio.file.Files
+import java.util.concurrent.ConcurrentHashMap
 import scala.collection.concurrent.TrieMap
+import scala.jdk.CollectionConverters._
 import scala.util.chaining._
 
 object CoderMaterializer {
   import com.spotify.scio.ScioContext
 
-  private[scio] case class CoderOptions(nullableCoders: Boolean, kryo: KryoOptions)
+  private[scio] case class CoderOptions(
+    nullableCoders: Boolean,
+    kryo: KryoOptions,
+    zstdDictMapping: Map[String, Array[Byte]]
+  )
   private[scio] object CoderOptions {
+    private val cache: ConcurrentHashMap[PipelineOptions, CoderOptions] = new ConcurrentHashMap()
```

Comment:
Are PipelineOptions implementations always guaranteed to be stable map keys? In Beam I don't see that any of them implement hashCode/equals. Maybe we could use the [appName](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/options/ApplicationNameOptions.java#L31) as a key instead? That should help keep the serialzed Coder byte size down too

Reply:
Ugh no, will revisit this